### DISCUSSION TITLE
Unable to follow docs to implement Aspire specific CosmosDB package

### DISCUSSION THREAD

**Discussion by swegele** [microsoft/aspire#2318] [General]
I'm loving the Aspire framework!  It's a lot to learn but I am getting it.

I have a working simple .NET 8 Aspire solution with 4 projects:
- WebAPI project with minimal API endpoint
- Application project (referenced by WebAPI project - assembly where MediatR stuff)
- CosmosClientService project (referenced by Application project - assembly where my CosmosClient is created as a singleton service)
- Common (with DTOs)

I am currently using the **Microsoft.Azure.Cosmos 3.38.0** package.

I spent almost a whole day trying to understand the instructions on the following page for Cosmos and Aspire:
https://learn.microsoft.com/en-us/dotnet/aspire/database/azure-cosmos-db-component
I tried every combination I could think of to get the aspire package **Aspire.Microsoft.Azure.Cosmos** to actually do something different than the normal cosmos package.  I was still able to connect to Cosmos with this aspire package, but nothing acted differently than before.  I didn't see any difference in traces or logs or anything.

I find the instructions confusing and unclear:

- Why do the instructions say to call this method three different times?  Wouldn't that make three different CosmosClient objects?
`builder.AddAzureCosmosDB("cosmosConnectionName");`
`var cosmosdb = builder.AddAzureCosmosDB("cdb");`
`builder.AddAzureCosmosDB("cosmosdb");`

- Why call `AddAzureCosmosDB` in the AppHost project?  That project does not need a `CosmosClient` object AFAIK.

- When successfully setup, how will I know?  What is it I am looking for to see success?

Are you able to show a simple example that is known to work where I can see the whole appsettings.json and Program.cs files.
Are you able to show a "before vs after" blurb about how one would know that the aspire comos package is doing it's thing?

Thanks for all your work on Aspire.

**Comment by JamesLMiller** [comment]
I asked a similar question [Aspire Emulator](https://github.com/dotnet/docs-aspire/issues/392). And was told that an example app was being worked on. I have been racking my brain here as well. I would like to use the emulator for testing, and have added the code in the right spots, but the missing link seems to be the actual connection to the emulator.

**Comment by davidfowl** [comment]
There’s confusion about method names here and what gets called from where. It’s consistent feedback that we’ve heard from customers.

In the AppHost project, you are wiring up resource. Aspire.Hosting.Azure has the cosmos db resource. This can be used to model the cosmos dependency for spinning up an emulator or a remote cosmos in azure. 

In the other projects you’re using the aspire component. As for how it’s different from using the underlying package:

- it injects client APIs in to the DI container with an appropriate lifetime. As well as configuring multiple clients with keyed DI
- It wires up telemetry to push traces and logs to otel
- It allows configuring the client from config 
- Handles both connection strings and managed identity without having to change the code 


You can get an overview of components here

https://github.com/dotnet/aspire/blob/main/src/Components/Aspire_Components_Progress.md


the key thing to remember is 

AppHost - Resources (launch projects, processes containers, cloud resources)
Service projects - components (libraries used at runtime)

**Reply by swegele** [reply]
EUREKA!!  Thank you @davidfowl  - Lightbulb went on when you pointed out the two method names are the same `AddAzureCosmosDB` but they are pointing to two different assemblies like so:

First one in AppHost project (for resource and config):
```
Aspire.Hosting.Azure.Data.Cosmos.AddAzureCosmosDB(...)
```

Second one in the project that gets the actual CosmosClient object for data access:
```
Aspire.Microsoft.Azure.Cosmos.AddAzureCosmosDB(...)
```

My second confusion/struggle was around the connection string naming and location.  System couldn't find connection strings without doing the `Name, ConnectionString` style.  
This worked and is from the AppHost Program.cs:
```
var cosmosEndpoint = builder.Configuration.GetConnectionString("cosmos_endpoint");

var my_cosmos_db = builder.AddAzureCosmosDB("cosmosEndpoint", cosmosEndpoint)
    .AddDatabase("my_cosmos_db");

builder.AddProject<Projects.MyWebAPI>("mywebapi")
    .WithReference(my_cosmos_db);

```

The appsettings.json file content for the AppHost project:
```
{
  "Logging": {
      ...
    }
  },
  "ConnectionStrings": {
    "cosmos_endpoint": "https://my-cosmos-account.documents.azure.com:443/"
  }
}
```

Finally here is the DependencyInjection.cs in MyApplication project (called from the Program.cs in MyWebAPI project):
```
        builder.AddAzureCosmosDB("my_cosmos_db", (AzureCosmosDBSettings settings) =>
        {
            settings.Credential = new DefaultAzureCredential();
        });

```


FYI @JamesLMiller - I did a test and everything works now so hopefully the above helps you too.

**Reply by JamesLMiller** [reply]
That does help. I guess the missing piece for me, is when I setup my context:

```
protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
{
    optionsBuilder.UseCosmos(
        "AccountEndpoint=https://localhost:8081/;AccountKey=C2y6yDjf5/R+ob0N8A7Cgv30VRDJIWEHLM+4QDU5DE2nQ9nDuVTqobD4b8mGGyPMbIZnqyMsEcaGQy67XIw/Jw==",
        databaseName: "cosmosDatabase",
        options =>
        {
            options.ConnectionMode(ConnectionMode.Gateway);

        });
}
```

It feels like I am duplicating work here adding endpoints and DB names again. There must be a way to connect to a provider without duplicating this setup. I thought that was the major bonus of Aspire.

**Comment by davidfowl** [comment]
> I asked a similar question https://github.com/dotnet/docs-aspire/issues/392. And was told that an example app was being worked on. I have been racking my brain here as well. I would like to use the emulator for testing, and have added the code in the right spots, but the missing link seems to be the actual connection to the emulator.

Here’s an example https://github.com/dotnet/aspire/blob/88406425f85daccec9b5cb3f22d5c8d07ebf3a8d/playground/CosmosEndToEnd/CosmosEndToEnd.AppHost/Program.cs#L6

**Reply by codebuzz** [reply]
The code mentioned by @swegele supposedly doesn't compile anymore and there doesn't seem to be any way to get the aspire cosmos package working. I always get failed to provision error (Azure.RequestFailedException: The access token is from the wrong issuer 'https://sts.windows.net/f8cdef31-a31e-4b4a-93e4-5f571e91255a/') in the aspire dashboard and tracing and metrics are simply not there.
